### PR TITLE
Add validation for field alias mappings.

### PR DESCRIPTION
**Pull request by jtibshirani**
The following restrictions are enforced on the target of an alias:
 * The target must be a concrete field, and not an object or another field alias.
 * The target field must exist at the time the alias is created.
 * If nested objects are defined, a field alias must have the same nested scope as its target.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java [hunk 1]
```diff
@@ -70,43 +70,71 @@ public FieldTypeLookup copyAndAddAll(String type,
...
+            List<String> conflicts = new ArrayList<>();
+            existingFieldType.checkCompatibility(newFieldType, conflicts);
+            if (conflicts.isEmpty() == false) {
+                throw new IllegalArgumentException("Mapper for [" + fieldName +
+                    "] conflicts with existing mapping:\n" + conflicts.toString());
+            }
+        }
+    }
+
+    private void validateAlias(String aliasName,
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java]: maybe some docs would help on the two above methods, for instance to say that new aliases have not been added yet when validateField is called and that new fields have already been added when validateAlias is called

File: server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class MapperMergeValidator {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java]: can it be pkg-private? please also add some minimal javadocs

File: server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                                                  Map<String, ObjectMapper> fullPathObjectMappers) {
+        for (FieldAliasMapper mapper : fieldAliasMappers) {
+            String aliasName = mapper.name();
+            String path = mapper.path();
+
+            String aliasScope = getNestedScope(aliasName, fullPathObjectMappers);
+            String pathScope = getNestedScope(path, fullPathObjectMappers);
+            if (!Objects.equals(aliasScope, pathScope)) {
+                throw new IllegalArgumentException("Invalid [path] value [" + path + "] for field alias [" +
+                    aliasName + "]: an alias must have the same nested scope as its target.");
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java]: maybe add what we computed as the scope of the alias and the path to the error message

**Review by jpountz — APPROVED** [review]
I left some minor comment but it looks good to me. Thanks for tackling this!

**Comment by jpountz** [comment]
@jtibshirani Could you update the description of this PR to describe what is being validated? This is important since this PR might be linked from release notes.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java (hunk 1 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java]: 👍

File: server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java (hunk 2 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java]: 👍

File: server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java (hunk 3 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java]: 👍

**Pull request merged**